### DISCUSSION TITLE
MongoDB start-up

### DISCUSSION THREAD

**Discussion by kaleming** [fastapi/fastapi#2721] [Questions] [answered]
Hi,

Supposing I have a simple main script app:

```
app = FastAPI(title="Async FastAPI")

app.include_router(student.router, tags=["Students"], prefix="/student")
```
if I add:

```
@app.on_event('startup')
async def startup():
    setup_mongodb(app)

```
or 

```
app.add_event_handler("startup", connect_to_mongo)
```

How should I call mongo client to run from another .py (not main.py) file which is responsible to perform CRUD function that should be connect to MongoDB.

Example:

How should I import db to use inside add_student function which is not located in main.py file ?

```
# Add a new student into to the database
async def add_student(student_data: dict) -> dict:
    student = await db.student_collection.insert_one(student_data)
    new_student = await student_collection.find_one({"_id": student.inserted_id})
    return student_helper(new_student)
```

**Comment by krishnardt** [comment]
You can create the database and collections on event startup, but can't start the database and also it is not safe process to keep the database connection open till you stop the server

You can create a new setup file for database and from there.. you can call your db connection through Depends parameter whenever you are deal with crud ops of database .

**Comment by kaleming** [comment, marked as answer]
Something like this ?

[main.py](https://github.com/michaldev/fastapi-async-mongodb/blob/master/app/main.py)
[crud.py](https://github.com/michaldev/fastapi-async-mongodb/blob/master/app/rest/posts.py)

Do you have a simple example of the recommended way to startup on main.py function ?

Should I use `@app.on_event('startup') `or` app.add_event_handler("startup", connect_to_mongo) `?

**Reply by krishnardt** [reply]
No, as per my knowledge, there is no way..

if get_my database in crud.py is a function that opens and closes the connecgtion, then there is not use of connect_to_db in startup. 
on event startup can be used to check all the services are working fine or starting and shutting down the protocols and other services.

to keep it simple,
you can start and stop protocols like rabbitMQ(AQMP protocols), HTTP protocols, wesocket using startup and shutdown.
but can't start the services like database.


but I have a question.
1.  you wanted to start the database from event startup or event handler and will be stopped on event shutdown, what if there is some database insertion failed and it raises the issue.. How would you  handle it?
2. why do you want to start the database from event startup?


below links will help you for better understanding..
[https://www.gitmemory.com/issue/tiangolo/fastapi/617/548368665](url)
[https://www.starlette.io/events/](url)
[https://stackoverflow.com/questions/61755472/event-handler-on-fastapi]()

**Comment by kaleming** [comment]
I was following the codes from testdriven.io and the creator of piccolo repository, and both uses this pattern:

[piccolo example](https://github.com/piccolo-orm/piccolo_examples/blob/master/headless_blog_fastapi/app.py)

[test driven example](https://github.com/testdrivenio/fastapi-crud-async/blob/master/src/app/main.py)

That's why I am still confused about it.

If I understand your point, you are saying that this @app.on_event('startup')  can't be used on crud.py. They have different purposes.
